src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,392 @@
+lemma coprime_prod_left
+  {I : Type*}
+  {x : ℕ} {s : I → ℕ} {t : finset I} :
+  (∀ (i : I), i ∈ t → nat.coprime (s i) x) → nat.coprime (∏ (i : I) in t, s i) x :=
+begin
+  intro h,
+  rw ← nat.is_coprime_iff_coprime,
+  have := @nat.cast_prod _ ℤ _ s t,
+  simp [ -nat.cast_prod ] at this,
+  rw this,
+  apply is_coprime.prod_left,
+  intros i hi,
+  rw nat.is_coprime_iff_coprime,
+  apply h i hi,
+end
+
+section with_group
+
+parameters {G : Type*} [group G]
+
+-- TODO: Move to suitable file
+@[simp]
+lemma order_of_inv (x : G) : order_of x⁻¹ = order_of x :=
+begin
+  apply nat.dvd_antisymm; rewrite order_of_dvd_iff_pow_eq_one,
+  { rw [inv_pow, pow_order_of_eq_one, one_inv], },
+  { nth_rewrite 0 ← (inv_inv x), rw [inv_pow, pow_order_of_eq_one, one_inv], }
+end
+
+-- TODO: Move to suitable file? Or too specialized?
+lemma mul_eq_one_of_disjoint
+  {H₁ H₂ : subgroup G} (hdis : disjoint H₁ H₂) {x y : G} (hx : x ∈ H₁) (hy : y ∈ H₂)
+  (heq : x * y = 1) : x = 1 ∧ y = 1 :=
+begin
+  have : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
+  subst this,
+  have hy := H₂.inv_mem_iff.mp hy,
+  have : x ∈ H₁ ⊓ H₂, by { simp, cc },
+  rw [hdis.eq_bot, subgroup.mem_bot] at this,
+  subst this,
+  simp
+end
+
+section family_of_groups
+
+-- We have an family of groups
+parameters {I : Type*} [fintype I] [decidable_eq I] {H : I → Type*} [∀ i, group (H i)]
+
+-- And morphism ϕ into G
+parameters (ϕ : Π (i : I), H i →* G)
+
+-- We assume that the elements of different morphism commute
+-- Since we need this all over the place we wrap it up in `fact`
+-- TODO: Worth making this a real `class` over `ϕ`?
+parameters [hcomm : fact (∀ (i j : I), i ≠ j → ∀ (x : H i) (y : H j), commute (ϕ i x) (ϕ j y))]
+include hcomm
+
+-- Elements of `Π (i : I), H i` are called `f` and `g` here
+variables (f g : Π (i : I), H i)
+
+namespace pi_hom_restr
+
+-- In this section, we restrict the hom to a set S
+variables (S : finset I)
+
+/-- The underlying function of `pi_hom_restr.hom` -/
+def to_fun (S : finset I) : G := finset.noncomm_prod S (λ i, ϕ i (f i)) $
+  by { rintros i - j -, by_cases (i = j), { subst h }, { exact hcomm.elim _ _ h _ _} }
+
+@[simp]
+lemma to_fun_empty : to_fun f ∅ = 1 := by simp [to_fun]
+
+@[simp]
+lemma to_fun_insert_of_not_mem (S : finset I) (i : I) (h : i ∉ S) :
+  to_fun f (insert i S) = ϕ i (f i) * to_fun f S :=
+finset.noncomm_prod_insert_of_not_mem _ _ _ _ h
+
+@[simp]
+lemma to_fun_one : to_fun 1 S = 1 :=
+begin
+   induction S using finset.cons_induction_on with i S hnmem ih,
+   { simp }, { simp [ih, hnmem], }
+end
+
+lemma to_fun_commutes (i : I) (hnmem : i ∉ S) :
+  commute (ϕ i (g i)) (to_fun f S) :=
+begin
+  induction S using finset.induction_on with j S hnmem' ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem'],
+
+    have hij : i ≠ j, by {simp at hnmem, tauto},
+    have hiS : i ∉ S, by {simp at hnmem, tauto},
+    calc ϕ i (g i) * (ϕ j (f j) * (to_fun ϕ f S : G)) -- TODO: Why do I have to mention `ϕ`?
```

Comment:
I thought by making `ϕ` a parameter I don’t have to mention it here again. So why do I?

Reply:
Ok, known problem, tacic mode undoes parameters: https://github.com/leanprover/lean/issues/1773

Reply:
```suggestion
    calc ϕ i (g i) * (ϕ j (f j) * (to_fun ϕ f S : G))
```